File: books/cs/introduction-to-algorithms/chapter6/Scala/heap_sort.scala
import scala.util.Random
import scala.collection._

object HeapSort{
    def main(args: Array[String]):Unit = {
        val dataSetSize = 10
        var srcdata = util.Random.shuffle(100 to 999).take(dataSetSize).toArray;
        var aaa = new Array[Int](dataSetSize)
        var bbb = new Array[Int](aaa.size)
        val digit = 3
        var idx:Int = 0
        var jdx:Int = 0

        println(s"the source array, ${srcdata.toList}")
        heap_sort(srcdata,dataSetSize)
        println(s"the sorted data, ${srcdata.toList}")
    }

    def swap(srcdata: Array[Int], p: Int, r: Int):Unit = {
        var temp = srcdata(p)
        srcdata(p) = srcdata(r)
        srcdata(r) = temp
    }

    def left( i: Int ): Int = {
        return(2*i)
    }
    def right( i: Int ): Int = {
        return(2*i+1)
    }


    def max_heapify(srcdata: Array[Int], len: Int, i: Int):Unit = {
        val l = left(i)
        val r = right(i)
        var largest = 0

        if( l<len && srcdata(l)>srcdata(i) ){
            largest = l
        }else largest = i
        if( r<len && srcdata(r)>srcdata(largest) ){
            largest = r
        }

        if( largest != i ){
            swap(srcdata,largest,i)
            max_heapify(srcdata,len,largest)
        }

    }

    def build_max_heap(srcdata: Array[Int], len: Int):Unit = {
        var idx: Int = 0

        for( idx <- len/2 to 0 by -1 ){
            max_heapify(srcdata,len,idx)
        }
    }

    def heap_sort(srcdata: Array[Int], len: Int):Unit = {
        var idx: Int = 0
        var datasize = len

        build_max_heap(srcdata,len)
        for( idx <- len-1 to 1 by -1 ){
            swap(srcdata,0,idx)
            datasize = datasize -1
            max_heapify(srcdata,datasize,0)
        }
    }

}

File: books/cs/general-purpose-algorithms/introduction-to-algorithms2/chapter6/R/heap_sort.R
#
# heap sort algorithm
# introduction to algorithm, 3rd edition, chapter 6
#
# clean the workspace and memory

rm( list=ls() )
gc()

aaa <- sample(999,10,replace = T)
srcdata <- aaa[aaa>100]

swap <- function(i,j){
  temp <- srcdata[i]
  srcdata[i] <<- srcdata[j]
  srcdata[j] <<- temp
}

left <- function(i){
  return(2*i)
}

right <- function(i){
  return(2*i+1)
}

max_heapify <- function(len,i){
  l <- left(i)
  r <- right(i)
  largest <- 0
  
  if( l<=len && srcdata[l]>srcdata[i] ){
    largest <- l
  }else largest <- i
  if( r<=len && srcdata[r]>srcdata[largest] ){
    largest <- r
  }
  
  if( largest != i ){
    swap(i,largest)
    max_heapify(len,largest)
  }
}

build_max_heap <- function(len){
  
  for( idx in floor(len/2):1 ){
	max_heapify(len,idx)
  }
}

heap_sort <- function(len){
  
  build_max_heap(len)
  for( idx in len:2 ){
	swap(idx,1)
	len <- len-1
	max_heapify(len,1)
  }
}

print(srcdata)
#quick_sort(1,length(srcdata))
heap_sort(length(srcdata))
print(srcdata)
